clang/test/SemaSYCL/sycl-restrict.cpp

```diff
@@ -36,7 +36,7 @@ void no_restriction(int p) {
   int index[p + 2];
 }
 void restriction(int p) {
-  int index[p + 2]; // expected-error {{variable length arrays are not supported for the current target}}
+  int index[p + 2]; // expected-error 2{{variable length arrays are not supported for the current target}}
```

Comment:
What are the two errors here?  How'd we get the extra one?  I'd like to understand how this happens.

Reply:
I have a theory .

Everything else in the `isa_B` function context has double the number of expectation values than you would expect.  I think this is because the calling hierarchy looks like this:
`main -> kernel -> usage -> kernel -> isa_B -> restriction`
Note that there is a kernel inside the `usage` function, which is called from a different kernel. 

The change in this PR moves the diagnostics for VLA to be a Deferred Diagnostic, like most of the others. The Deferred Diagnostics walk the final produced kernel looking to see if there were any warnings deferred earlier for that thing, and then they emit.  

So my theory is the the old system, using the AST Visitor, only visited it once (and then issued an immediate diagnostic). But with this change ,the deferred diagnostic system essentially visits it twice, because of the nested kernels.

I did a quick test and move that call to `restriction()` up a level to `usage()` , and now it fires just once, not twice.

Reply:
I can make that change to the sycl-restrict, to make it clearer. I do not know why there are nested kernels for these tests. Let me know if you would like a change.

Reply:
Right, I just want to make sure its still an appropriate amount. sycl-restrict is criminally oversized unfortunately, so understanding this stuff is annoying.  If there is a simplification that you can makes it clearer, I would be appreciative.

Reply:
I there any reason for having nested kernels in these tests? Or is it just an artifact of this files history?

I'd be happy to get rid of the kernel nesting which would tidy things up a bit. But I don't want to undo anything important.

Otherwise, I can simply move the testing of the VLA .    My own inclination is to get rid of the nested kernels entirely.

Reply:
I think the nesting is to make sure that nesting doesn't break the diagnostics, which I think it did at one point.  IMO, just some sort of separation to the test, or some way of associating which diagnostic goes with which 'using' tree would be helpful.

How about putting the notes associated with each diagnostic right next to the diagnostic, so that I can follow them?

Reply:
I can't say why we somehow get these "nested kernels" in tests.
I don't think that "nested kernels" is completely valid SYCL code, but technically I can't call it completely invalid, because SYCL doesn't define what is `sycl_kernel` attribute and what is happening when some function with such attribute is happening it is only our implementation detail.
`sycl_kernel` attribute should be used in SYCL headers implementation and I don't think that someone would do "nested kernels" here and recalling what is actually happening with functions with `sycl_kernel` attribute I think we even don't produce invalid SPIR-V when do "nested kernels".
So, "nested kernels" is a strange phenomenon, but I don't mind having it in the tests. Because compiler tests shouldn't be ideal and simple code. We develop a compiler, user can do any weird stuff and we shouldn't crash, at least.

Reply:
you mean the `expected-note` ? 
In the cases here, the `Check_VLA_Restriction` namespace names what it's testing pretty well, but the parents ( `isa_B` and `usage` ) are called for multiple diagnostics, so it's not a 1:1 relationship.

Or am I misunderstanding your intent?

Reply:
I'm saying just move the notes around so it is clear what they are coming from.  In this case, I see that the VLA throws two errors, and presumably a bunch of notes, but I can't follow what the include-stack of notes is, so I don't know if they are valid.

Reply:
I can add more comments to explain, but I can't move the note expectations themselves.  
I have tried putting the lit errors and notes together in the past, but it doesn't seem like that's how it works. The note is pointing to the original caller kernel/lambda , and that's what the expectation keys off.   These "called by" notes are  also one of the reasons we bring these mock kernels into the test, even though we have dummy kernels in our common mock sycl header.  Because here we need to add "called by"  expectations to it.

Reply:
Of course you can move them around.  Use the "@" syntax for the expected-errors.

Reply:
Sorry to be dense.  Like this? 
```suggestion
  // expected-note@71 2{{called by 'isa_B'}}
  int index[p + 2]; // expected-error 2{{variable length arrays are not supported for the current target}}
```
Is there another way of using the "@" syntax so it's not absolute line numbers? I have used the relative "@+1" to put these immediately before a line, but in the case we are discussing, the caller we are mentioning is far away.   This approach seems like it'll just break the minute someone edits the file. 
Is this -verify expectation system documented anywhere? I've searched, but the llvm-lit docs don't talk about this part.  Are there other ways of using the "@" in the expectation?

Reply:
See https://clang.llvm.org/doxygen/classclang_1_1VerifyDiagnosticConsumer.html

Reply:
thanks. A veritable rosetta stone, that. Do you have the decipher key for Linear A as well?

Reply:
Negative ghostrider.

Reply:
Ok, I have pushed added markers and moved the notes into reasonable places for the VLA test, which is the one test this PR touches.  

I would like to just go ahead and finish that and move _all_ the stray note expectations in sycl-restrict.cpp and put them close to their caller.  Is that ok to do in this PR? Or should I open a different one for that?